### PR TITLE
feat(benchmarks): Write better benchmark for testing the aggregator

### LINKED ISSUE #3844: Improve benchmarks for the aggregator

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR reworks the benchmarks for the aggregator by adding more fine-grained control over certain parameters.

Closes: https://github.com/getsentry/relay/issues/3844

#skip-changelog

### CODE REVIEW

File: relay-metrics/benches/benches.rs [hunk 1]
```diff
@@ -138,21 +200,22 @@ fn bench_insert_and_flush(c: &mut Criterion) {
...
+                        for (project_key, bucket) in input.get_buckets(timestamp) {
                             aggregator.merge(project_key, bucket, None).unwrap();
                         }
                         aggregator
                     },
                     |mut aggregator| {
                         // XXX: Ideally we'd want to test the entire try_flush here, but spawning
                         // a service is too much work here.
-                        aggregator.pop_flush_buckets(false);
+                        black_box(aggregator.pop_flush_buckets(false));
```
Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: Nice!

File: relay-metrics/benches/benches.rs [hunk 2]
```diff
@@ -1,135 +1,197 @@
...
+        Self {
+            min: range.start,
+            max: range.end,
+            current_value: RefCell::new(0),
         }
+    }
+
+    fn next(&self) -> usize {
+        let seed = *self.current_value.borrow() as u128;
+        let mut generator = Pcg32::new((seed >> 64) as u64, seed as u64);
```
Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: Can we use `SmallRng::from_seed` here instead? (less dependencies)

Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: You create the prng only once in the constructor then keep calling it, no need to change the seed everytime.

File: relay-metrics/benches/benches.rs [hunk 3]
```diff
@@ -1,135 +1,197 @@
...
+    metric_ids_generator: NumbersGenerator,
+    project_keys_generator: NumbersGenerator,
+    timestamp_shifts_generator: NumbersGenerator,
+}
+impl BucketsGenerator {
+    fn get_buckets(&self, base_timestamp: UnixTimestamp) -> Vec<(ProjectKey, Bucket)> {
+        let mut buckets = Vec::with_capacity(self.num_buckets);
+
+        let backdated = ((self.num_buckets as f32 * self.percentage_backdated) as usize)
+            .clamp(0, self.num_buckets);
```
Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: It's a `usize` already, so you can just do `.min(self.num_buckets)`

File: relay-metrics/benches/benches.rs (hunk 2 above)
Comment (iambriccardo) [reply relay-metrics/benches/benches.rs]: Does it spit always a different number deterministically if we do so?

File: relay-metrics/benches/benches.rs [hunk 4]
```diff
@@ -1,135 +1,197 @@
...
+                metric_ids_generator: NumbersGenerator::new(1..1_000_000),
+                project_keys_generator: NumbersGenerator::new(1..1_000_000),
+                timestamp_shifts_generator: NumbersGenerator::new(10..50),
+            },
+        ),
     ];
 
-    for input in &inputs {
+    for (input_name, input) in &inputs {
         c.bench_with_input(
```
Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: Can make a criterion group here and specify throughput, compare to here https://github.com/getsentry/relay/blob/4d5941afc9544700e6a0b04cff861cf56455cb84/relay-cardinality/benches/redis_impl.rs#L198

File: relay-metrics/benches/benches.rs [hunk 5]
```diff
@@ -1,135 +1,197 @@
...
+                        let timestamp = UnixTimestamp::now();
                         let aggregator: Aggregator = Aggregator::new(config.clone());
-                        (aggregator, input.get_buckets())
+                        (aggregator, input.get_buckets(timestamp))
                     },
                     |(mut aggregator, buckets)| {
                         for (project_key, bucket) in buckets {
-                            aggregator.merge(project_key, bucket, None).unwrap();
+                            #[allow(clippy::unit_arg)]
+                            black_box(aggregator.merge(project_key, bucket, None).unwrap());
```
Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: I think the black box should be on the aggregator after the loop.

File: relay-metrics/benches/benches.rs (hunk 4 above)
Comment (iambriccardo) [reply relay-metrics/benches/benches.rs]: Oh nice!

File: relay-metrics/benches/benches.rs (hunk 5 above)
Comment (iambriccardo) [reply relay-metrics/benches/benches.rs]: I thought that `black_box` is handled specifically by the compiler just by wrapping an expression with it. Maybe I need to better read the docs.

Comment (iambriccardo) [reply relay-metrics/benches/benches.rs:194]: Okay, read the docs, you are right. It actually tries to be pessimistic about the dummy value.

Comment (iambriccardo) [reply relay-metrics/benches/benches.rs]: I read the docs, it's indeed doing some stuff with the result value, treating it pessimistically.

Comment (Dav1dde) [reply relay-metrics/benches/benches.rs]: The result value here is just the unit type. I am not sure what the compiler is allowed to optimize here, but for me it makes more sense on the "result".

File: relay-metrics/Cargo.toml [hunk 6]
```diff
@@ -22,6 +22,7 @@ fnv = { workspace = true }
 hash32 = { workspace = true }
 hashbrown = { workspace = true }
 itertools = { workspace = true }
+rand = { workspace = true }
```
Comment (Dav1dde) [inline relay-metrics/Cargo.toml]: That's a `dev-dependency`.

File: relay-metrics/benches/benches.rs [hunk 7]
```diff
@@ -1,158 +1,231 @@
...
                         }
                     },
                     BatchSize::SmallInput,
                 )
             },
         );
 
-        c.bench_with_input(
-            BenchmarkId::new("bench_flush_metrics", input),
+        group.bench_with_input(
```
Comment (Dav1dde) [inline relay-metrics/benches/benches.rs]: Shouldn't be the same group, this is a different benchmark with a different throughput! (no for loop in the body).

**Review by Dav1dde — APPROVED** [review]

**Pull request merged**